mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,789 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements a loop interchange pass that optimizes for locality
+// (spatial and temporal - both self and group) and parallelism for multicores,
+// to minimize the frequency of synchronization. The pass works for both
+// perfectly nested and imperfectly nested loops (any level of nesting). However
+// in the presence of affine.if statements and/or non-rectangular iteration
+// space, the pass simply bails out - leaving the original loop nest unchanged.
+// The pass is triggered by the command line flag -affine-loop-interchange.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Analysis/Utils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Passes.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include <algorithm>
+#include <cmath>
+#include <numeric>
+using namespace mlir;
+namespace {
+struct LoopInterchange : public AffineLoopInterchangeBase<LoopInterchange> {
+  void runOnFunction() override;
+  void handleImperfectlyNestedAffineLoops(Operation &funcOp);
+  void runOnAffineLoopNest();
+
+private:
+  /// Default cache line size(in bytes). Useful for getting a measure of the
+  /// locality of each loop in a given loop nest.
+  constexpr static unsigned kCacheLineSize = 64;
+
+  /// Default element size to be used if a memref does not have a static shape.
+  constexpr static unsigned kDefaultEltSize = 8;
+
+  bool isRectangularAffineForLoopNest();
+
+  void getLoopCarriedDependenceVector();
+
+  void getAllLoadStores();
+
+  void getCacheLineAccessCounts(
+      DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+          &loopAccessMatrices,
+      DenseMap<Operation *, unsigned> &elementSizes);
+
+  uint64_t getNumCacheLinesSpatialReuse(ArrayRef<unsigned> perm);
+
+  uint64_t getNumSyncs(ArrayRef<unsigned> perm);
+
+  uint64_t getNumCacheLinesTemporalReuse(
+      ArrayRef<unsigned> permutation,
+      DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+          &loopAccessMatrices,
+      uint64_t maxTemporalReuse);
+
+  bool getBestPermutation(DenseMap<Value, unsigned> &loopIndexMap,
+                          SmallVector<unsigned, 4> &bestPerm);
+
+  // Loop Carried Dependence vector. A 'true' at index 'i' means that the loop
+  // at depth 'i' carries a dependence.
+  SmallVector<bool, 4> loopCarriedDV;
+
+  // Iteration count of each loop in the loop nest.
+  SmallVector<unsigned, 4> loopIterationCounts;
+
+  // The loop nest.
+  SmallVector<AffineForOp, 4> loopVector;
+
+  /// Number of cache lines accessed by each loop in the loop nest.
+  DenseMap<const AffineForOp *, uint64_t> cacheLinesAccessCounts;
```

Comment:
These would be wrong. The `AffineForOp` is just a value type wrapper around the actual op. The same `Operation` could thus potentially have different `AffineForOp *`. You need to hash on `Operation *`. This also means you aren't getting any caching here.